### PR TITLE
utils/exec polling fixes

### PR DESCRIPTION
**Pull request by tbzatek**
Well, this essentially ended up as a complete rewrite of `_utils_exec_and_report_progress()`. Tried to cover various corner cases, heavy testing is required.

Fixes: #236

### CODE REVIEW

**Comment by vojtechtrefny** [comment]
Looks good to me in general, but it also seems to break all F2FS test cases on Debian. `dump.f2fs` prints some nullbytes to stdout on Debian so I guess the nulbytes handling doesn't work as expected in the new version.

**Review by vojtechtrefny — CHANGES_REQUESTED** [review]
Looks good to me in general, but it also seems to break all F2FS test cases on Debian. dump.f2fs prints some nullbytes to stdout on Debian so I guess the nulbytes handling doesn't work as expected in the new version.

**Comment by tbzatek** [comment]
> Looks good to me in general, but it also seems to break all F2FS test cases on Debian. dump.f2fs prints some nullbytes to stdout on Debian so I guess the nulbytes handling doesn't work as expected in the new version.

Fixed, tests seems to pass now.

**Review by vojtechtrefny — APPROVED** [review]
Looks good to me.

**Comment by vojtechtrefny** [comment]
This also fixes #236 Mostly by removing all the GIOChannel code, but still one less clang issue for us :-)

**Comment by vojtechtrefny** [comment]
There is one new issue from coverity, not sure what coverity really means by tainted data here.

```
Error: TAINTED_SCALAR (CWE-20): [#def7]
libblockdev-2.99/src/utils/exec.c:295: tainted_argument: Calling function "read" taints argument "*buf". [Note: The source code implementation of the function has been overridden by a builtin model.]
libblockdev-2.99/src/utils/exec.c:296: tainted_data_transitive: Call to function "g_string_append_len" with tainted argument "buf" transitively taints "*read_buffer".
libblockdev-2.99/src/utils/exec.c:300: lower_bounds: Checking lower bounds of unsigned scalar "read_buffer->len" by taking the true branch of "read_buffer->len > *read_buffer_pos".
libblockdev-2.99/src/utils/exec.c:301: tainted_data: Passing tainted expression "read_buffer->len - *read_buffer_pos" to "bd_strchr_len_null", which uses it as an offset.
libblockdev-2.99/src/utils/exec.c:301: remediation: Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.
#  299|           /* process the fresh data by lines */
#  300|           if (read_buffer->len > *read_buffer_pos) {
#  301|->             while ((newline_pos = bd_strchr_len_null (read_buffer->str + *read_buffer_pos, read_buffer->len - *read_buffer_pos, '\n'))) {
#  302|                   line = g_strndup (read_buffer->str + *read_buffer_pos, newline_pos - read_buffer->str - *read_buffer_pos + 1);
#  303|                   if (prog_extract && prog_extract (line, progress))
```

**Comment by tbzatek** [comment]
> There is one new issue from coverity, not sure what coverity really means by tainted data here.

Perhaps just a friendly warning about possible buffer overflow or read beyond the buffer. I believe this is properly checked through the `read()` return values.

**Comment by vojtechtrefny** [comment]
> > There is one new issue from coverity, not sure what coverity really means by tainted data here.
> 
> Perhaps just a friendly warning about possible buffer overflow or read beyond the buffer. I believe this is properly checked through the `read()` return values.

Ok. Do you want to investigate this further or should we just ignore it for now (or possibly add the coverity ignore comment to the code)?

**Comment by tbzatek** [comment]
> Ok. Do you want to investigate this further or should we just ignore it for now (or possibly add the coverity ignore comment to the code)?

I took one more look but still think we're safe here - the `while` cycle only considers return values greater than zero and `read()` guarantees return value equal or lower of the size of the buffer. I don't see a problem here and have no idea what Coverity expects.

Let's see what @vpodzime would find.

**Comment by vpodzime** [comment]
I'll review this over the weekend :crossed_fingers:

File: src/utils/exec.c [hunk 1]
```diff
@@ -348,190 +423,128 @@ static gboolean _utils_exec_and_report_progress (const gchar **argv, const BDExt
...
-            if (error && (*error))
-                g_prefix_error (error, "Failed to write to stdin of the process: ");
-            else
-                g_set_error (error, BD_UTILS_EXEC_ERROR, BD_UTILS_EXEC_ERROR_FAILED,
-                             "Failed to write to stdin of the process.");
-
+        buf_written = write (in_fd, input, strlen (input));
+        if (buf_written < (ssize_t) strlen (input)) {
+            g_set_error (error, BD_UTILS_EXEC_ERROR, BD_UTILS_EXEC_ERROR_FAILED,
+                         "Failed to write to stdin of the process: %m");
```
Comment (vpodzime) [inline src/utils/exec.c:438]: `write()` doesn't necessarily have to write all the data at once. This needs a loop making sure all bytes where written or there was an actual error.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. And I don't see a problem with how `read()` is done.

File: src/utils/exec.h [hunk 2]
```diff
@@ -21,10 +21,30 @@ typedef void (*BDUtilsProgFunc) (guint64 task_id, BDUtilsProgStatus status, guin
...
  *
- * Returns: whether the line was a progress reporting line or not
+ * Callback function used to process a line captured from spawned command's standard
+ * output and standard error output. Typically used to extract completion percentage
+ * of a long-running job.
+ *
+ * Note that both outputs are read simultaneously with no guarantees of message order
+ * this function is called with.
+ *
+ * The value the @completion points to may contain value previously returned from
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/utils/exec.h:34]: Hmm, I think this is correct, _"The value the completion points to"_ ... another option could be _"points at"_?

File: src/utils/exec.c [hunk 3]
```diff
@@ -264,6 +265,81 @@ gboolean bd_utils_exec_and_report_status_error (const gchar **argv, const BDExtr
...
+
+    ret = memchr (haystack, needle, haystack_len);
+    ret_null = memchr (haystack, 0, haystack_len);
+    if (ret && ret_null)
+        return MIN (ret, ret_null);
+    else
+        return MAX (ret, ret_null);
+}
+
+static gboolean _process_fd_event (gint fd, struct pollfd *poll_fd, GString *read_buffer, GString *filtered_buffer, gsize *read_buffer_pos, gboolean *done, guint64 progress_id, guint8 *progress, BDUtilsProgExtract prog_extract, GError **error) {
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/utils/exec.c:285]: Done.

File: src/utils/exec.h [hunk 4]
```diff
@@ -21,10 +21,30 @@ typedef void (*BDUtilsProgFunc) (guint64 task_id, BDUtilsProgStatus status, guin
...
+ * this callback or a zero when called for the first time. This is useful for
+ * extractors where only a some kind of tick mark is printed out as a progress.
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/utils/exec.h:36]: Clarified. Thanks for being a consumer-tester :-)

File: src/utils/exec.c [hunk 5]
```diff
@@ -264,6 +265,81 @@ gboolean bd_utils_exec_and_report_status_error (const gchar **argv, const BDExtr
...
+    if (! *done && (poll_fd->revents & POLLIN)) {
+        /* read until we get EOF (0) or error (-1), expecting EAGAIN */
+        while ((num_read = read (fd, buf, _EXEC_BUF_SIZE)) > 0)
+            g_string_append_len (read_buffer, buf, num_read);
+        errno_saved = errno;
+
+        /* process the fresh data by lines */
+        if (read_buffer->len > *read_buffer_pos) {
+            while ((newline_pos = bd_strchr_len_null (read_buffer->str + *read_buffer_pos, read_buffer->len - *read_buffer_pos, '\n'))) {
+                line = g_strndup (read_buffer->str + *read_buffer_pos, newline_pos - read_buffer->str - *read_buffer_pos + 1);
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/utils/exec.c:302]: Yeah, it's misleading on a first sight. Changed.

File: src/utils/exec.c (hunk 1 above)
Comment (tbzatek) [reply src/utils/exec.c:438]: Good point! Added a cycle and some more checks.

**Comment by tbzatek** [comment]
> Looks good to me otherwise. And I don't see a problem with how `read()` is done.

Thanks. Fixed and rebased. Also added some input tests as this was never tested here (perhaps in other units though).

**Review by vojtechtrefny — APPROVED** [review]
Still looks good to me, but you should fix the CLANG_WARNING from csmock.

**Review by vpodzime — APPROVED** [review]
ACK
